py/asmxtensa.c

```diff
@@ -42,6 +42,10 @@
 #define ET_OUT_OF_RANGE MP_ERROR_TEXT("ERROR: xtensa %q out of range")
 #define ET_NOT_ALIGNED MP_ERROR_TEXT("ERROR: %q %q not word-aligned")
 
+// A7 is part of the incoming parameter registers set in both CALL0 and
+// windowed ABIs, thus is not callee-saved and can be safely clobbered.
+#define REG_TEMP ASM_XTENSA_REG_A7
```

Comment:
This is problematic: A7 is used as a base pointer to the function table `mp_fun_table` when in windowed mode.  So clobbering it will lead to a crash if the function table is used afterwards.

Eg the following crashes on esp32:
```py
@micropython.native
def f(p: ptr32):
    p[256] = 1
    print('done')

f(bytearray(2048))
```
It's a bit tricky to fix.  The native emitter does allocate a register for calculating the index `reg_index` but that's not passed through to the backend emitter.

Reply:
This is weird, as I'm quite sure I ran the full set of `tests/micropython/` tests on a D0WD.  `tests/micropython/viper_ptr32_store_boundary.py` would have crashed when using the last set of offsets (around the 12-bits boundary mark).

But I've seen the constant in question so this needs fixing.  

I believe that using `REG_TEMP2` could be a temporary workaround - don't have an ESP32 available at the moment to test it.

`REG_TEMP2` is only used for implementing the `rot_three` opcode, and the native implementation whilst performing register stores would only use the stack for with that register.  How big can a single emitted function stack grow?  If that number is below 1024 bytes then `REG_TEMP2` should be it, otherwise I'll see how this can be worked around differently.

Reply:
> `tests/micropython/viper_ptr32_store_boundary.py` would have crashed when using the last set of offsets (around the 12-bits boundary mark).

The reason it doesn't crash is because it doesn't access A7 after clobbering it.  In my test above you need the `print('done')` to trigger the crash, because that line requires A7.

Reply:
OK, using `REG_TEMP2` (or the same register anyway) seems to work on a regular ESP32 board when running the test suite with `--via-mpy --emit native`.

I've had to extend the range of some branch opcodes to make `basics/frozenset_binop.py`, `basics/set_binop.py`, `extmod/framebuf1.py`, and `extmod/json_dumps_separators.py` actually compile for xtensa/xtensawin in the first place.

I haven't tested this on an ESP8266 yet (need to fetch the board from somewhere else) but, unless checking for `N_XTENSAWIN` when picking which register number to use is not correct, in this case if it works on ESP32 then it should also work for ESP8266.

Reply:
> I've had to extend the range of some branch opcodes to make `basics/frozenset_binop.py`, `basics/set_binop.py`, `extmod/framebuf1.py`, and `extmod/json_dumps_separators.py` actually compile for xtensa/xtensawin in the first place.

Excellent, thanks for that.  I had noticed for some time that these tests fail, but with all the other native changes in flight, I was waiting before mentioning it.  See related #17293.

Reply:
> I haven't tested this on an ESP8266 yet (need to fetch the board from somewhere else) but, unless checking for `N_XTENSAWIN` when picking which register number to use is not correct, in this case if it works on ESP32 then it should also work for ESP8266 unless this throws alignment off for inlineasm tests if I didn't handle things right when updating branch opcodes.

I've just had the chance to test this on an ESP8266 - inline asm branches aren't affected, some test crash with `--via-mpy --emit=native`.  I'm looking into it.

Reply:
OK, I've misread the ISA document - `BEQZ.N` and `BNEZ.N` can only operate with an *unsigned* displacement, and on an ESP32 there were no backward jumps below the (signed) 6-bits threshold that triggered the emission of those opcodes.

This should be fixed now!